lib/matplotlib/axis.py

```diff
@@ -1626,14 +1631,24 @@ def set_ticklabels(self, ticklabels, *, minor=False, **kwargs):
         """
         ticklabels = [t.get_text() if hasattr(t, 'get_text') else t
                       for t in ticklabels]
+        locator = (self.get_minor_locator() if minor
+                   else self.get_major_locator())
+        if isinstance(locator, mticker.FixedLocator):
+            tickd = {loc: lab for loc, lab in zip(locator.locs, ticklabels)}
+            formatter = mticker.FuncFormatter(self._format_with_dict, tickd)
```

Comment:
I'm not sure whether we want or not to make this extension to the FuncFormatter API (perhaps?) but if not, I think you can get away with
```
@staticmethod
def _format_with_ticks(tickd, x, pos): ...  # move tickd to first

...

formatter = FuncFormatter(functools.partial(self._format_with_ticks, tickd))
```
as the `partial` object will stay picklable (I think this is more or less the canonical way to bind some arguments while keeping picklability).